Lite playground design changes

aliabd opened the pull request:
Makes a lot of changes to the lite playground element: 

![Screen Shot 2024-03-24 at 4 45 57 AM](https://github.com/gradio-app/gradio/assets/9021060/87805237-1d9b-40b7-a1ca-df6d6bbf67bf)

- Adds an optional layout attribute to the custom element. If horizontal code and preview will always be side by side, if vertical they'll always be on top of each other. If it isn't provided it will switch at the md breakpoint. 
- Adds copy and download buttons to the code editor
- Fixes issue with CMD+Enter adding a new line 
- Cleans up the header, colors and font sizes. Moved things around a bit and removed the barrier in the middle. Not sure if the run button is in the right place though.. 

Having an issue getting the heights to work right, will spend more time on that.

To test just run 

```
sh scripts/build_lite.sh
python -m http.server --directory js/lite
```
and open `http://localhost:8000/`. you can also edit `js/lite/index.html`

whitphx commented on js/app/src/lite/Playground.svelte:
```diff
@@ -60,37 +60,55 @@
 		}
 	}
 
+	window.addEventListener("keydown", shortcut_run, true);
```
I think this event listener should be attached to each `<Code />` element, instead of the global window.

With the current setting, it behaves like below and I think it's not intuitive.
* `CMD+Enter` triggers a new run even if the editor is not focused.
* When there are multiple Playground elements, a single `CMD+Enter` triggers a new run on all the Playgrounds.

whitphx commented on js/app/src/lite/Playground.svelte:
```diff
@@ -213,9 +238,12 @@
 
 	.code-header {
 		padding-top: 0.25rem;
```
This `padding-top` should be removed?
The "app.py" text is vertically aligned to the "Run" text.
<table>
<tr>
<th>With it</th>
<th>Without it</th>
</tr>
<tr>
<td>
<img src="https://github.com/gradio-app/gradio/assets/3135397/0b265c34-bb31-4c03-a1c2-2e1da322e8f5" />
</td>
<td>
<img src="https://github.com/gradio-app/gradio/assets/3135397/a070f7c1-9df0-46af-96eb-86be33a7a1dc" />
</td>
</tr>
</table>

whitphx commented:
I also found when the system is set as dark-mode, the Gradio Playground is initialized in light-mode during the initial load, but becomes dark after the initialization finishes.
![CleanShot 2024-03-25 at 16 26 20](https://github.com/gradio-app/gradio/assets/3135397/fafe5a9b-d04d-491c-a9bc-8627a18d7007)

abidlabs commented:
Thanks @aliabd! UI/UX looks great. I don't have much to add beyond what @whitphx pointed out in his review. The only other thing I noticed was that copy button was not working for me. Clicking on it has no effect (does not copy anything to clipboard, no console error logs)

Also the padding around the code buttons maybe should be a bit smaller:

<img width="281" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/557c0e39-1b01-428a-a7db-8bb08b8e4eff">

Otherwise very nice!

aliabd commented:
@abidlabs navigator clipboard access is blocked by chrome on localhost. But I tested it and it will work fine otherwise. Will fix the padding

abidlabs commented:
Ah okay thanks for clarifying @aliabd! Btw I realized we don’t document the playground tag anywhere! We should add it to the Lite Guide?

aliabd commented:
@abidlabs yup will add to the guide as well. Btw, it's prettier that keeps removing the line breaks in index.html, will try to fix that

aliabd commented:
@abidlabs @whitphx this is ready for another pass. I fixed everything you pointed out. 

Also added dark mode! 

https://github.com/gradio-app/gradio/assets/9021060/69b8c9ef-3ed0-4bcb-9f62-a9f05723358d

abidlabs reviewed (approved):
LGTM @aliabd! Dark mode is slick!

My only UX feedback is that I wish that the code editor did not lose focus when I did CMD + Enter (minor but I think it would be an improvement as it would allow me to keep typing and easily seeing the preview)

abidlabs commented:
Hmm wait actually I'm seeing something weird -- when I open `http://localhost:8000/`, my memory usage jumps ~15GB and other tabs start to freeze up. Do you see this behavior?

aliabd commented:
@abidlabs i guess this is what happens when you have three gradio-lite elements on one page.. adding shared-worker reduces the memory load though. Just added it to index.html so at least the problem will be gone there. And we use it on the website already (that's why it doesn't reload lite for every page). 

I agree with you losing focus is annoying, but we can't fix that now because currently running the code reloads the whole playground element. we discussed this a bit here: https://github.com/gradio-app/gradio/pull/7660#discussion_r1520786298 but i think we should revisit later.

whitphx commented on js/app/src/lite/Playground.svelte:
```diff
@@ -60,65 +60,94 @@
+	let parent_container: HTMLDivElement;
+
+	onMount(() => {
+		var code_editors = document.getElementsByClassName("code-editor");
+		for (var i = 0; i < code_editors.length; i++) {
+			code_editors[i].addEventListener(
+				"keydown",
+				shortcut_run as EventListener,
+				true
+			);
+		}
```
This still has a problem that the shortcut key pressed on a single editor will trigger re-running all the mounted playgrounds on the page.
The event listener here should be attached only to the child `<Code />` in this component.

BTW, such "Enter+CMD" shortcut can be `<Code />` component's feature (maybe a new event dispatcher like `run_requested`?), WDYT @pngwn 

I think it can be a separate PR.

whitphx commented:
@aliabd Thank you, I checked the changes and confirmed it has nice look and feel!
I left one comment. I think it's technically a bug but not critical so we can deal with it later in a separate PR if this PR should go.

abidlabs commented:
> @abidlabs i guess this is what happens when you have three gradio-lite elements on one page.. adding shared-worker reduces the memory load though. Just added it to index.html so at least the problem will be gone there. And we use it on the website already (that's why it doesn't reload lite for every page).

Got it, thanks for adding this. I tested and confirmed that the memory usage is a lot better now. Can we just make this the default? Is there any downside of using shared-worker by default @whitphx @aliabd?

> I agree with you losing focus is annoying, but we can't fix that now because currently running the code reloads the whole playground element. we discussed this a bit here: https://github.com/gradio-app/gradio/pull/7660#discussion_r1520786298 but i think we should revisit later.

I agree sounds good.

Everything else lgtm!

abidlabs commented on guides/08_gradio-clients-and-lite/gradio-lite.md:
```diff
@@ -148,6 +148,21 @@ demo.launch()
 
+### Code and Demo Playground 
+
+If you'd like to see the code side-by-side with the demo just pass in the `playground` attribute to the gradio-lite element. This will create an interactive playground that allows you to change the code and update the demo! If you're using playground, you can also set layout to either 'vertical' or 'horizontal' which will determine if the code editor and preview are side-by-side or on top of each ...
```
> layout to either 'vertical' or 'horizontal'

Unclear how to do this, an example would help. Also there are other attributes, like shared-worker, right? Would be good to document them all

whitphx commented:
@abidlabs 
> Can we just make this the default? Is there any downside of using shared-worker by default?

Yea, that's an option.
On the other hand, these are what I thought when I made the SharedWorker mode optional, not the default:
* The worker logs are printed in the SharedWorker's console, not the normal developer console. So it's a bit cumbersome for users to get the logs to report errors. In Chrome, for example, users have to open chrome://inspect/#workers to access the SharedWorker's console.
* The SharedWorker mode has some things to know like below, so I thought users should be aware of using the SharedWorker mode by explicitly enabling it.
    * The file system is shared among all the Lite apps using the SharedWorker while each has a separated home dirs, so file operations outside the home e.g. on `/` may conflict.
    * Likewise, the Python runtime is shared, so apps may conflict. For example, if one app monkey-patches some module, another app using that module is affected too.

WDYT?

abidlabs commented:
Thanks for the info, I agree with keeping shared workers optional then. But let’s add this information to the documentation and make the recommendation that shared workers should be enabled if you have multiple playground instances on a webpage.

whitphx commented:
Yes, definitely.
Will write it.

whitphx commented:
Added a section about SharedWorker: #7923

aliabd commented:
Thanks @whitphx for the changes. I also edited the guide to make layout attribute clearer. 

Will merge this, we can make any other changes in a separate PR.
